<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 2.0.12">
<meta name="description" content="TransPi – a comprehensive TRanscriptome ANalysiS PIpeline for de novo transcriptome assembly">
<meta name="keywords" content="TransPi, transcriptome, assembly, annotation, Nextflow, pipeline">
<meta name="author" content="Ramón Rivera-Vicéns">
<title>TransPi Manual</title>
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:300,300italic,400,400italic,600,600italic%7CNoto+Serif:400,400italic,700,700italic%7CDroid+Sans+Mono:400,700">
<style>
/* Asciidoctor default stylesheet | MIT License | https://asciidoctor.org */
/* Uncomment @import statement to use as custom stylesheet */
/*@import "https://fonts.googleapis.com/css?family=Open+Sans:300,300italic,400,400italic,600,600italic%7CNoto+Serif:400,400italic,700,700italic%7CDroid+Sans+Mono:400,700";*/
article,aside,details,figcaption,figure,footer,header,hgroup,main,nav,section{display:block}
audio,video{display:inline-block}
audio:not([controls]){display:none;height:0}
html{font-family:sans-serif;-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%}
a{background:none}
a:focus{outline:thin dotted}
a:active,a:hover{outline:0}
h1{font-size:2em;margin:.67em 0}
abbr[title]{border-bottom:1px dotted}
b,strong{font-weight:bold}
dfn{font-style:italic}
hr{-moz-box-sizing:content-box;box-sizing:content-box;height:0}
mark{background:#ff0;color:#000}
code,kbd,pre,samp{font-family:monospace;font-size:1em}
pre{white-space:pre-wrap}
q{quotes:"\201C" "\201D" "\2018" "\2019"}
small{font-size:80%}
sub,sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}
sup{top:-.5em}
sub{bottom:-.25em}
img{border:0}
svg:not(:root){overflow:hidden}
figure{margin:0}
fieldset{border:1px solid silver;margin:0 2px;padding:.35em .625em .75em}
legend{border:0;padding:0}
button,input,select,textarea{font-family:inherit;font-size:100%;margin:0}
button,input{line-height:normal}
button,select{text-transform:none}
button,html input[type="button"],input[type="reset"],input[type="submit"]{-webkit-appearance:button;cursor:pointer}
button[disabled],html input[disabled]{cursor:default}
input[type="checkbox"],input[type="radio"]{box-sizing:border-box;padding:0}
button::-moz-focus-inner,input::-moz-focus-inner{border:0;padding:0}
textarea{overflow:auto;vertical-align:top}
table{border-collapse:collapse;border-spacing:0}
*,*::before,*::after{-moz-box-sizing:border-box;-webkit-box-sizing:border-box;box-sizing:border-box}
html,body{font-size:100%}
body{background:#fff;color:rgba(0,0,0,.8);padding:0;margin:0;font-family:"Noto Serif","DejaVu Serif",serif;font-weight:400;font-style:normal;line-height:1;position:relative;cursor:auto;tab-size:4;word-wrap:anywhere;-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased}
a:hover{cursor:pointer}
img,object,embed{max-width:100%;height:auto}
object,embed{height:100%}
img{-ms-interpolation-mode:bicubic}
.left{float:left!important}
.right{float:right!important}
.text-left{text-align:left!important}
.text-right{text-align:right!important}
.text-center{text-align:center!important}
.text-justify{text-align:justify!important}
.hide{display:none}
img,object,svg{display:inline-block;vertical-align:middle}
textarea{height:auto;min-height:50px}
select{width:100%}
.subheader,.admonitionblock td.content>.title,.audioblock>.title,.exampleblock>.title,.imageblock>.title,.listingblock>.title,.literalblock>.title,.stemblock>.title,.openblock>.title,.paragraph>.title,.quoteblock>.title,table.tableblock>.title,.verseblock>.title,.videoblock>.title,.dlist>.title,.olist>.title,.ulist>.title,.qlist>.title,.hdlist>.title{line-height:1.45;color:#7a2518;font-weight:400;margin-top:0;margin-bottom:.25em}
div,dl,dt,dd,ul,ol,li,h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6,pre,form,p,blockquote,th,td{margin:0;padding:0}
a{color:#2156a5;text-decoration:underline;line-height:inherit}
a:hover,a:focus{color:#1d4b8f}
a img{border:0}
p{font-family:inherit;font-weight:400;font-size:1em;line-height:1.6;margin-bottom:1.25em;text-rendering:optimizeLegibility}
p aside{font-size:.875em;line-height:1.35;font-style:italic}
h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{font-family:"Open Sans","DejaVu Sans",sans-serif;font-weight:300;font-style:normal;color:#ba3925;text-rendering:optimizeLegibility;margin-top:1em;margin-bottom:.5em;line-height:1.0125em}
h1 small,h2 small,h3 small,#toctitle small,.sidebarblock>.content>.title small,h4 small,h5 small,h6 small{font-size:60%;color:#e99b8f;line-height:0}
h1{font-size:2.125em}
h2{font-size:1.6875em}
h3,#toctitle,.sidebarblock>.content>.title{font-size:1.375em}
h4,h5{font-size:1.125em}
h6{font-size:1em}
hr{border:solid #dddddf;border-width:1px 0 0;clear:both;margin:1.25em 0 1.1875em;height:0}
em,i{font-style:italic;line-height:inherit}
strong,b{font-weight:bold;line-height:inherit}
small{font-size:60%;line-height:inherit}
code{font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;font-weight:400;color:rgba(0,0,0,.9)}
ul,ol,dl{font-size:1em;line-height:1.6;margin-bottom:1.25em;list-style-position:outside;font-family:inherit}
ul,ol{margin-left:1.5em}
ul li ul,ul li ol{margin-left:1.25em;margin-bottom:0;font-size:1em}
ul.square li ul,ul.circle li ul,ul.disc li ul{list-style:inherit}
ul.square{list-style-type:square}
ul.circle{list-style-type:circle}
ul.disc{list-style-type:disc}
ol li ul,ol li ol{margin-left:1.25em;margin-bottom:0}
dl dt{margin-bottom:.3125em;font-weight:bold}
dl dd{margin-bottom:1.25em}
abbr,acronym{text-transform:uppercase;font-size:90%;color:rgba(0,0,0,.8);border-bottom:1px dotted #ddd;cursor:help}
abbr{text-transform:none}
blockquote{margin:0 0 1.25em;padding:.5625em 1.25em 0 1.1875em;border-left:1px solid #ddd}
blockquote cite{display:block;font-size:.9375em;color:rgba(0,0,0,.6)}
blockquote cite::before{content:"\2014 \0020"}
blockquote cite a,blockquote cite a:visited{color:rgba(0,0,0,.6)}
blockquote,blockquote p{line-height:1.6;color:rgba(0,0,0,.85)}
@media screen and (min-width:768px){h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{line-height:1.2}
h1{font-size:2.75em}
h2{font-size:2.3125em}
h3,#toctitle,.sidebarblock>.content>.title{font-size:1.6875em}
h4{font-size:1.4375em}}
table{background:#fff;margin-bottom:1.25em;border:solid 1px #dedede;word-wrap:normal}
table thead,table tfoot{background:#f7f8f7}
table thead tr th,table thead tr td,table tfoot tr th,table tfoot tr td{padding:.5em .625em .625em;font-size:inherit;color:rgba(0,0,0,.8);text-align:left}
table tr th,table tr td{padding:.5625em .625em;font-size:inherit;color:rgba(0,0,0,.8)}
table tr.even,table tr.alt{background:#f8f8f7}
table thead tr th,table tfoot tr th,table tbody tr td,table tr td,table tfoot tr td{line-height:1.6}
h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{line-height:1.2;word-spacing:-.05em}
h1 strong,h2 strong,h3 strong,#toctitle strong,.sidebarblock>.content>.title strong,h4 strong,h5 strong,h6 strong{font-weight:400}
.center{margin-left:auto;margin-right:auto}
.stretch{width:100%}
.clearfix::before,.clearfix::after,.float-group::before,.float-group::after{content:" ";display:table}
.clearfix::after,.float-group::after{clear:both}
:not(pre).nobreak{word-wrap:normal}
:not(pre).nowrap{white-space:nowrap}
:not(pre).pre-wrap{white-space:pre-wrap}
:not(pre):not([class^=L])>code{font-size:.9375em;font-style:normal!important;letter-spacing:0;padding:.1em .5ex;word-spacing:-.15em;background:#f7f7f8;-webkit-border-radius:4px;border-radius:4px;line-height:1.45;text-rendering:optimizeSpeed}
pre{color:rgba(0,0,0,.9);font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;line-height:1.45;text-rendering:optimizeSpeed}
pre code,pre pre{color:inherit;font-size:inherit;line-height:inherit}
pre>code{display:block}
pre.nowrap,pre.nowrap pre{white-space:pre;word-wrap:normal}
em em{font-style:normal}
strong strong{font-weight:400}
.keyseq{color:rgba(51,51,51,.8)}
kbd{font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;display:inline-block;color:rgba(0,0,0,.8);font-size:.65em;line-height:1.45;background:#f7f7f7;border:1px solid #ccc;-webkit-border-radius:3px;border-radius:3px;-webkit-box-shadow:0 1px 0 rgba(0,0,0,.2),0 0 0 .1em white inset;box-shadow:0 1px 0 rgba(0,0,0,.2),0 0 0 .1em #fff inset;margin:0 .15em;padding:.2em .5em;vertical-align:middle;position:relative;top:-.1em;white-space:nowrap}
.keyseq kbd:first-child{margin-left:0}
.keyseq kbd:last-child{margin-right:0}
.menuseq,.menuref{color:#000}
.menuseq b:not(.caret),.menuref{font-weight:inherit}
.menuseq{word-spacing:-.02em}
.menuseq b.caret{font-size:1.25em;line-height:.8}
.menuseq i.caret{font-weight:bold;text-align:center;width:.45em}
b.button::before,b.button::after{position:relative;top:-1px;font-weight:400}
b.button::before{content:"[";padding:0 3px 0 2px}
b.button::after{content:"]";padding:0 2px 0 3px}
p a>code:hover{color:rgba(0,0,0,.9)}
#header,#content,#footnotes,#footer{width:100%;margin-left:auto;margin-right:auto;margin-top:0;margin-bottom:0;max-width:62.5em;*zoom:1;position:relative;padding-left:.9375em;padding-right:.9375em}
#header::before,#header::after,#content::before,#content::after,#footnotes::before,#footnotes::after,#footer::before,#footer::after{content:" ";display:table}
#header::after,#content::after,#footnotes::after,#footer::after{clear:both}
#content{margin-top:1.25em}
#content::before{content:none}
#header>h1:first-child{color:rgba(0,0,0,.85);margin-top:2.25rem;margin-bottom:0}
#header>h1:first-child+#toc{margin-top:8px;border-top:1px solid #dddddf}
#header>h1:only-child,body.toc2 #header>h1:nth-last-child(2){border-bottom:1px solid #dddddf;padding-bottom:8px}
#header .details{border-bottom:1px solid #dddddf;line-height:1.45;padding-top:.25em;padding-bottom:.25em;padding-left:.25em;color:rgba(0,0,0,.6);display:-ms-flexbox;display:-webkit-flex;display:flex;-ms-flex-flow:row wrap;-webkit-flex-flow:row wrap;flex-flow:row wrap}
#header .details span:first-child{margin-left:-.125em}
#header .details span.email a{color:rgba(0,0,0,.85)}
#header .details br{display:none}
#header .details br+span::before{content:"\00a0\2013\00a0"}
#header .details br+span.author::before{content:"\00a0\22c5\00a0";color:rgba(0,0,0,.85)}
#header .details br+span#revremark::before{content:"\00a0|\00a0"}
#header #revnumber{text-transform:capitalize}
#header #revnumber::after{content:"\00a0"}
#content>h1:first-child:not([class]){color:rgba(0,0,0,.85);border-bottom:1px solid #dddddf;padding-bottom:8px;margin-top:0;padding-top:1rem;margin-bottom:1.25rem}
#toc{border-bottom:1px solid #e7e7e9;padding-bottom:.5em}
#toc>ul{margin-left:.125em}
#toc ul.sectlevel0>li>a{font-style:italic}
#toc ul.sectlevel0 ul.sectlevel1{margin:.5em 0}
#toc ul{font-family:"Open Sans","DejaVu Sans",sans-serif;list-style-type:none}
#toc li{line-height:1.3334;margin-top:.3334em}
#toc a{text-decoration:none}
#toc a:active{text-decoration:underline}
#toctitle{color:#7a2518;font-size:1.2em}
@media screen and (min-width:768px){#toctitle{font-size:1.375em}
body.toc2{padding-left:15em;padding-right:0}
#toc.toc2{margin-top:0!important;background:#f8f8f7;position:fixed;width:15em;left:0;top:0;border-right:1px solid #e7e7e9;border-top-width:0!important;border-bottom-width:0!important;z-index:1000;padding:1.25em 1em;height:100%;overflow:auto}
#toc.toc2 #toctitle{margin-top:0;margin-bottom:.8rem;font-size:1.2em}
#toc.toc2>ul{font-size:.9em;margin-bottom:0}
#toc.toc2 ul ul{margin-left:0;padding-left:1em}
#toc.toc2 ul.sectlevel0 ul.sectlevel1{padding-left:0;margin-top:.5em;margin-bottom:.5em}
body.toc2.toc-right{padding-left:0;padding-right:15em}
body.toc2.toc-right #toc.toc2{border-right-width:0;border-left:1px solid #e7e7e9;left:auto;right:0}}
@media screen and (min-width:1280px){body.toc2{padding-left:20em;padding-right:0}
#toc.toc2{width:20em}
#toc.toc2 #toctitle{font-size:1.375em}
#toc.toc2>ul{font-size:.95em}
#toc.toc2 ul ul{padding-left:1.25em}
body.toc2.toc-right{padding-left:0;padding-right:20em}}
#content #toc{border-style:solid;border-width:1px;border-color:#e0e0dc;margin-bottom:1.25em;padding:1.25em;background:#f8f8f7;-webkit-border-radius:4px;border-radius:4px}
#content #toc>:first-child{margin-top:0}
#content #toc>:last-child{margin-bottom:0}
#footer{max-width:none;background:rgba(0,0,0,.8);padding:1.25em}
#footer-text{color:rgba(255,255,255,.8);line-height:1.44}
#content{margin-bottom:.625em}
.sect1{padding-bottom:.625em}
@media screen and (min-width:768px){#content{margin-bottom:1.25em}
.sect1{padding-bottom:1.25em}}
.sect1:last-child{padding-bottom:0}
.sect1+.sect1{border-top:1px solid #e7e7e9}
#content h1>a.anchor,h2>a.anchor,h3>a.anchor,#toctitle>a.anchor,.sidebarblock>.content>.title>a.anchor,h4>a.anchor,h5>a.anchor,h6>a.anchor{position:absolute;z-index:1001;width:1.5ex;margin-left:-1.5ex;display:block;text-decoration:none!important;visibility:hidden;text-align:center;font-weight:400}
#content h1>a.anchor::before,h2>a.anchor::before,h3>a.anchor::before,#toctitle>a.anchor::before,.sidebarblock>.content>.title>a.anchor::before,h4>a.anchor::before,h5>a.anchor::before,h6>a.anchor::before{content:"\00A7";font-size:.85em;display:block;padding-top:.1em}
#content h1:hover>a.anchor,#content h1>a.anchor:hover,h2:hover>a.anchor,h2>a.anchor:hover,h3:hover>a.anchor,#toctitle:hover>a.anchor,.sidebarblock>.content>.title:hover>a.anchor,h3>a.anchor:hover,#toctitle>a.anchor:hover,.sidebarblock>.content>.title>a.anchor:hover,h4:hover>a.anchor,h4>a.anchor:hover,h5:hover>a.anchor,h5>a.anchor:hover,h6:hover>a.anchor,h6>a.anchor:hover{visibility:visible}
#content h1>a.link,h2>a.link,h3>a.link,#toctitle>a.link,.sidebarblock>.content>.title>a.link,h4>a.link,h5>a.link,h6>a.link{color:#ba3925;text-decoration:none}
#content h1>a.link:hover,h2>a.link:hover,h3>a.link:hover,#toctitle>a.link:hover,.sidebarblock>.content>.title>a.link:hover,h4>a.link:hover,h5>a.link:hover,h6>a.link:hover{color:#a53221}
details,.audioblock,.imageblock,.literalblock,.listingblock,.stemblock,.videoblock{margin-bottom:1.25em}
details>summary:first-of-type{cursor:pointer;display:list-item;outline:none;margin-bottom:.75em}
.admonitionblock td.content>.title,.audioblock>.title,.exampleblock>.title,.imageblock>.title,.listingblock>.title,.literalblock>.title,.stemblock>.title,.openblock>.title,.paragraph>.title,.quoteblock>.title,table.tableblock>.title,.verseblock>.title,.videoblock>.title,.dlist>.title,.olist>.title,.ulist>.title,.qlist>.title,.hdlist>.title{text-rendering:optimizeLegibility;text-align:left;font-family:"Noto Serif","DejaVu Serif",serif;font-size:1rem;font-style:italic}
table.tableblock.fit-content>caption.title{white-space:nowrap;width:0}
.paragraph.lead>p,#preamble>.sectionbody>[class="paragraph"]:first-of-type p{font-size:1.21875em;line-height:1.6;color:rgba(0,0,0,.85)}
table.tableblock #preamble>.sectionbody>[class="paragraph"]:first-of-type p{font-size:inherit}
.admonitionblock>table{border-collapse:separate;border:0;background:none;width:100%}
.admonitionblock>table td.icon{text-align:center;width:80px}
.admonitionblock>table td.icon img{max-width:none}
.admonitionblock>table td.icon .title{font-weight:bold;font-family:"Open Sans","DejaVu Sans",sans-serif;text-transform:uppercase}
.admonitionblock>table td.content{padding-left:1.125em;padding-right:1.25em;border-left:1px solid #dddddf;color:rgba(0,0,0,.6);word-wrap:anywhere}
.admonitionblock>table td.content>:last-child>:last-child{margin-bottom:0}
.exampleblock>.content{border-style:solid;border-width:1px;border-color:#e6e6e6;margin-bottom:1.25em;padding:1.25em;background:#fff;-webkit-border-radius:4px;border-radius:4px}
.exampleblock>.content>:first-child{margin-top:0}
.exampleblock>.content>:last-child{margin-bottom:0}
.sidebarblock{border-style:solid;border-width:1px;border-color:#dbdbd6;margin-bottom:1.25em;padding:1.25em;background:#f3f3f2;-webkit-border-radius:4px;border-radius:4px}
.sidebarblock>:first-child{margin-top:0}
.sidebarblock>:last-child{margin-bottom:0}
.sidebarblock>.content>.title{color:#7a2518;margin-top:0;text-align:center}
.exampleblock>.content>:last-child>:last-child,.exampleblock>.content .olist>ol>li:last-child>:last-child,.exampleblock>.content .ulist>ul>li:last-child>:last-child,.exampleblock>.content .qlist>ol>li:last-child>:last-child,.sidebarblock>.content>:last-child>:last-child,.sidebarblock>.content .olist>ol>li:last-child>:last-child,.sidebarblock>.content .ulist>ul>li:last-child>:last-child,.sidebarblock>.content .qlist>ol>li:last-child>:last-child{margin-bottom:0}
.literalblock pre,.listingblock>.content>pre{-webkit-border-radius:4px;border-radius:4px;overflow-x:auto;padding:1em;font-size:.8125em}
@media screen and (min-width:768px){.literalblock pre,.listingblock>.content>pre{font-size:.90625em}}
@media screen and (min-width:1280px){.literalblock pre,.listingblock>.content>pre{font-size:1em}}
.literalblock pre,.listingblock>.content>pre:not(.highlight),.listingblock>.content>pre[class="highlight"],.listingblock>.content>pre[class^="highlight "]{background:#f7f7f8}
.literalblock.output pre{color:#f7f7f8;background:rgba(0,0,0,.9)}
.listingblock>.content{position:relative}
.listingblock code[data-lang]::before{display:none;content:attr(data-lang);position:absolute;font-size:.75em;top:.425rem;right:.5rem;line-height:1;text-transform:uppercase;color:inherit;opacity:.5}
.listingblock:hover code[data-lang]::before{display:block}
.listingblock.terminal pre .command::before{content:attr(data-prompt);padding-right:.5em;color:inherit;opacity:.5}
.listingblock.terminal pre .command:not([data-prompt])::before{content:"$"}
.listingblock pre.highlightjs{padding:0}
.listingblock pre.highlightjs>code{padding:1em;-webkit-border-radius:4px;border-radius:4px}
.listingblock pre.prettyprint{border-width:0}
.prettyprint{background:#f7f7f8}
pre.prettyprint .linenums{line-height:1.45;margin-left:2em}
pre.prettyprint li{background:none;list-style-type:inherit;padding-left:0}
pre.prettyprint li code[data-lang]::before{opacity:1}
pre.prettyprint li:not(:first-child) code[data-lang]::before{display:none}
table.linenotable{border-collapse:separate;border:0;margin-bottom:0;background:none}
table.linenotable td[class]{color:inherit;vertical-align:top;padding:0;line-height:inherit;white-space:normal}
table.linenotable td.code{padding-left:.75em}
table.linenotable td.linenos{border-right:1px solid currentColor;opacity:.35;padding-right:.5em}
pre.pygments .lineno{border-right:1px solid currentColor;opacity:.35;display:inline-block;margin-right:.75em}
pre.pygments .lineno::before{content:"";margin-right:-.125em}
.quoteblock{margin:0 1em 1.25em 1.5em;display:table}
.quoteblock:not(.excerpt)>.title{margin-left:-1.5em;margin-bottom:.75em}
.quoteblock blockquote,.quoteblock p{color:rgba(0,0,0,.85);font-size:1.15rem;line-height:1.75;word-spacing:.1em;letter-spacing:0;font-style:italic;text-align:justify}
.quoteblock blockquote{margin:0;padding:0;border:0}
.quoteblock blockquote::before{content:"\201c";float:left;font-size:2.75em;font-weight:bold;line-height:.6em;margin-left:-.6em;color:#7a2518;text-shadow:0 1px 2px rgba(0,0,0,.1)}
.quoteblock blockquote>.paragraph:last-child p{margin-bottom:0}
.quoteblock .attribution{margin-top:.75em;margin-right:.5ex;text-align:right}
.verseblock{margin:0 1em 1.25em}
.verseblock pre{font-family:"Open Sans","DejaVu Sans",sans;font-size:1.15rem;color:rgba(0,0,0,.85);font-weight:300;text-rendering:optimizeLegibility}
.verseblock pre strong{font-weight:400}
.verseblock .attribution{margin-top:1.25rem;margin-left:.5ex}
.quoteblock .attribution,.verseblock .attribution{font-size:.9375em;line-height:1.45;font-style:italic}
.quoteblock .attribution br,.verseblock .attribution br{display:none}
.quoteblock .attribution cite,.verseblock .attribution cite{display:block;letter-spacing:-.025em;color:rgba(0,0,0,.6)}
.quoteblock.abstract blockquote::before,.quoteblock.excerpt blockquote::before,.quoteblock .quoteblock blockquote::before{display:none}
.quoteblock.abstract blockquote,.quoteblock.abstract p,.quoteblock.excerpt blockquote,.quoteblock.excerpt p,.quoteblock .quoteblock blockquote,.quoteblock .quoteblock p{line-height:1.6;word-spacing:0}
.quoteblock.abstract{margin:0 1em 1.25em;display:block}
.quoteblock.abstract>.title{margin:0 0 .375em;font-size:1.15em;text-align:center}
.quoteblock.excerpt>blockquote,.quoteblock .quoteblock{padding:0 0 .25em 1em;border-left:.25em solid #dddddf}
.quoteblock.excerpt,.quoteblock .quoteblock{margin-left:0}
.quoteblock.excerpt blockquote,.quoteblock.excerpt p,.quoteblock .quoteblock blockquote,.quoteblock .quoteblock p{color:inherit;font-size:1.0625rem}
.quoteblock.excerpt .attribution,.quoteblock .quoteblock .attribution{color:inherit;text-align:left;margin-right:0}
p.tableblock:last-child{margin-bottom:0}
td.tableblock>.content{margin-bottom:1.25em;word-wrap:anywhere}
td.tableblock>.content>:last-child{margin-bottom:-1.25em}
table.tableblock,th.tableblock,td.tableblock{border:0 solid #dedede}
table.grid-all>*>tr>*{border-width:1px}
table.grid-cols>*>tr>*{border-width:0 1px}
table.grid-rows>*>tr>*{border-width:1px 0}
table.frame-all{border-width:1px}
table.frame-ends{border-width:1px 0}
table.frame-sides{border-width:0 1px}
table.frame-none>colgroup+*>:first-child>*,table.frame-sides>colgroup+*>:first-child>*{border-top-width:0}
table.frame-none>:last-child>:last-child>*,table.frame-sides>:last-child>:last-child>*{border-bottom-width:0}
table.frame-none>*>tr>:first-child,table.frame-ends>*>tr>:first-child{border-left-width:0}
table.frame-none>*>tr>:last-child,table.frame-ends>*>tr>:last-child{border-right-width:0}
table.stripes-all tr,table.stripes-odd tr:nth-of-type(odd),table.stripes-even tr:nth-of-type(even),table.stripes-hover tr:hover{background:#f8f8f7}
th.halign-left,td.halign-left{text-align:left}
th.halign-right,td.halign-right{text-align:right}
th.halign-center,td.halign-center{text-align:center}
th.valign-top,td.valign-top{vertical-align:top}
th.valign-bottom,td.valign-bottom{vertical-align:bottom}
th.valign-middle,td.valign-middle{vertical-align:middle}
table thead th,table tfoot th{font-weight:bold}
tbody tr th{background:#f7f8f7}
tbody tr th,tbody tr th p,tfoot tr th,tfoot tr th p{color:rgba(0,0,0,.8);font-weight:bold}
p.tableblock>code:only-child{background:none;padding:0}
p.tableblock{font-size:1em}
ol{margin-left:1.75em}
ul li ol{margin-left:1.5em}
dl dd{margin-left:1.125em}
dl dd:last-child,dl dd:last-child>:last-child{margin-bottom:0}
ol>li p,ul>li p,ul dd,ol dd,.olist .olist,.ulist .ulist,.ulist .olist,.olist .ulist{margin-bottom:.625em}
ul.checklist,ul.none,ol.none,ul.no-bullet,ol.no-bullet,ol.unnumbered,ul.unstyled,ol.unstyled{list-style-type:none}
ul.no-bullet,ol.no-bullet,ol.unnumbered{margin-left:.625em}
ul.unstyled,ol.unstyled{margin-left:0}
ul.checklist{margin-left:.625em}
ul.checklist li>p:first-child>.fa-square-o:first-child,ul.checklist li>p:first-child>.fa-check-square-o:first-child{width:1.25em;font-size:.8em;position:relative;bottom:.125em}
ul.checklist li>p:first-child>input[type="checkbox"]:first-child{margin-right:.25em}
ul.inline{display:-ms-flexbox;display:-webkit-box;display:flex;-ms-flex-flow:row wrap;-webkit-flex-flow:row wrap;flex-flow:row wrap;list-style:none;margin:0 0 .625em -1.25em}
ul.inline>li{margin-left:1.25em}
.unstyled dl dt{font-weight:400;font-style:normal}
ol.arabic{list-style-type:decimal}
ol.decimal{list-style-type:decimal-leading-zero}
ol.loweralpha{list-style-type:lower-alpha}
ol.upperalpha{list-style-type:upper-alpha}
ol.lowerroman{list-style-type:lower-roman}
ol.upperroman{list-style-type:upper-roman}
ol.lowergreek{list-style-type:lower-greek}
.hdlist>table,.colist>table{border:0;background:none}
.hdlist>table>tbody>tr,.colist>table>tbody>tr{background:none}
td.hdlist1,td.hdlist2{vertical-align:top;padding:0 .625em}
td.hdlist1{font-weight:bold;padding-bottom:1.25em}
td.hdlist2{word-wrap:anywhere}
.literalblock+.colist,.listingblock+.colist{margin-top:-.5em}
.colist td:not([class]):first-child{padding:.4em .75em 0;line-height:1;vertical-align:top}
.colist td:not([class]):first-child img{max-width:none}
.colist td:not([class]):last-child{padding:.25em 0}
.thumb,.th{line-height:0;display:inline-block;border:solid 4px #fff;-webkit-box-shadow:0 0 0 1px #ddd;box-shadow:0 0 0 1px #ddd}
.imageblock.left{margin:.25em .625em 1.25em 0}
.imageblock.right{margin:.25em 0 1.25em .625em}
.imageblock>.title{margin-bottom:0}
.imageblock.thumb,.imageblock.th{border-width:6px}
.imageblock.thumb>.title,.imageblock.th>.title{padding:0 .125em}
.image.left,.image.right{margin-top:.25em;margin-bottom:.25em;display:inline-block;line-height:0}
.image.left{margin-right:.625em}
.image.right{margin-left:.625em}
a.image{text-decoration:none;display:inline-block}
a.image object{pointer-events:none}
sup.footnote,sup.footnoteref{font-size:.875em;position:static;vertical-align:super}
sup.footnote a,sup.footnoteref a{text-decoration:none}
sup.footnote a:active,sup.footnoteref a:active{text-decoration:underline}
#footnotes{padding-top:.75em;padding-bottom:.75em;margin-bottom:.625em}
#footnotes hr{width:20%;min-width:6.25em;margin:-.25em 0 .75em;border-width:1px 0 0}
#footnotes .footnote{padding:0 .375em 0 .225em;line-height:1.3334;font-size:.875em;margin-left:1.2em;margin-bottom:.2em}
#footnotes .footnote a:first-of-type{font-weight:bold;text-decoration:none;margin-left:-1.05em}
#footnotes .footnote:last-of-type{margin-bottom:0}
#content #footnotes{margin-top:-.625em;margin-bottom:0;padding:.75em 0}
.gist .file-data>table{border:0;background:#fff;width:100%;margin-bottom:0}
.gist .file-data>table td.line-data{width:99%}
div.unbreakable{page-break-inside:avoid}
.big{font-size:larger}
.small{font-size:smaller}
.underline{text-decoration:underline}
.overline{text-decoration:overline}
.line-through{text-decoration:line-through}
.aqua{color:#00bfbf}
.aqua-background{background:#00fafa}
.black{color:#000}
.black-background{background:#000}
.blue{color:#0000bf}
.blue-background{background:#0000fa}
.fuchsia{color:#bf00bf}
.fuchsia-background{background:#fa00fa}
.gray{color:#606060}
.gray-background{background:#7d7d7d}
.green{color:#006000}
.green-background{background:#007d00}
.lime{color:#00bf00}
.lime-background{background:#00fa00}
.maroon{color:#600000}
.maroon-background{background:#7d0000}
.navy{color:#000060}
.navy-background{background:#00007d}
.olive{color:#606000}
.olive-background{background:#7d7d00}
.purple{color:#600060}
.purple-background{background:#7d007d}
.red{color:#bf0000}
.red-background{background:#fa0000}
.silver{color:#909090}
.silver-background{background:#bcbcbc}
.teal{color:#006060}
.teal-background{background:#007d7d}
.white{color:#bfbfbf}
.white-background{background:#fafafa}
.yellow{color:#bfbf00}
.yellow-background{background:#fafa00}
span.icon>.fa{cursor:default}
a span.icon>.fa{cursor:inherit}
.admonitionblock td.icon [class^="fa icon-"]{font-size:2.5em;text-shadow:1px 1px 2px rgba(0,0,0,.5);cursor:default}
.admonitionblock td.icon .icon-note::before{content:"\f05a";color:#19407c}
.admonitionblock td.icon .icon-tip::before{content:"\f0eb";text-shadow:1px 1px 2px rgba(155,155,0,.8);color:#111}
.admonitionblock td.icon .icon-warning::before{content:"\f071";color:#bf6900}
.admonitionblock td.icon .icon-caution::before{content:"\f06d";color:#bf3400}
.admonitionblock td.icon .icon-important::before{content:"\f06a";color:#bf0000}
.conum[data-value]{display:inline-block;color:#fff!important;background:rgba(0,0,0,.8);-webkit-border-radius:50%;border-radius:50%;text-align:center;font-size:.75em;width:1.67em;height:1.67em;line-height:1.67em;font-family:"Open Sans","DejaVu Sans",sans-serif;font-style:normal;font-weight:bold}
.conum[data-value] *{color:#fff!important}
.conum[data-value]+b{display:none}
.conum[data-value]::after{content:attr(data-value)}
pre .conum[data-value]{position:relative;top:-.125em}
b.conum *{color:inherit!important}
.conum:not([data-value]):empty{display:none}
dt,th.tableblock,td.content,div.footnote{text-rendering:optimizeLegibility}
h1,h2,p,td.content,span.alt{letter-spacing:-.01em}
p strong,td.content strong,div.footnote strong{letter-spacing:-.005em}
p,blockquote,dt,td.content,span.alt{font-size:1.0625rem}
p{margin-bottom:1.25rem}
.sidebarblock p,.sidebarblock dt,.sidebarblock td.content,p.tableblock{font-size:1em}
.exampleblock>.content{background:#fffef7;border-color:#e0e0dc;-webkit-box-shadow:0 1px 4px #e0e0dc;box-shadow:0 1px 4px #e0e0dc}
.print-only{display:none!important}
@page{margin:1.25cm .75cm}
@media print{*{-webkit-box-shadow:none!important;box-shadow:none!important;text-shadow:none!important}
html{font-size:80%}
a{color:inherit!important;text-decoration:underline!important}
a.bare,a[href^="#"],a[href^="mailto:"]{text-decoration:none!important}
a[href^="http:"]:not(.bare)::after,a[href^="https:"]:not(.bare)::after{content:"(" attr(href) ")";display:inline-block;font-size:.875em;padding-left:.25em}
abbr[title]::after{content:" (" attr(title) ")"}
pre,blockquote,tr,img,object,svg{page-break-inside:avoid}
thead{display:table-header-group}
svg{max-width:100%}
p,blockquote,dt,td.content{font-size:1em;orphans:3;widows:3}
h2,h3,#toctitle,.sidebarblock>.content>.title{page-break-after:avoid}
#header,#content,#footnotes,#footer{max-width:none}
#toc,.sidebarblock,.exampleblock>.content{background:none!important}
#toc{border-bottom:1px solid #dddddf!important;padding-bottom:0!important}
body.book #header{text-align:center}
body.book #header>h1:first-child{border:0!important;margin:2.5em 0 1em}
body.book #header .details{border:0!important;display:block;padding:0!important}
body.book #header .details span:first-child{margin-left:0!important}
body.book #header .details br{display:block}
body.book #header .details br+span::before{content:none!important}
body.book #toc{border:0!important;text-align:left!important;padding:0!important;margin:0!important}
body.book #toc,body.book #preamble,body.book h1.sect0,body.book .sect1>h2{page-break-before:always}
.listingblock code[data-lang]::before{display:block}
#footer{padding:0 .9375em}
.hide-on-print{display:none!important}
.print-only{display:block!important}
.hide-for-print{display:none!important}
.show-for-print{display:inherit!important}}
@media print,amzn-kf8{#header>h1:first-child{margin-top:1.25rem}
.sect1{padding:0!important}
.sect1+.sect1{border:0}
#footer{background:none}
#footer-text{color:rgba(0,0,0,.6);font-size:.9em}}
@media amzn-kf8{#header,#content,#footnotes,#footer{padding:0}}
</style>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
<style>
/* Stylesheet for CodeRay to match GitHub theme | MIT License | http://foundation.zurb.com */
pre.CodeRay{background:#f7f7f8}
.CodeRay .line-numbers{border-right:1px solid currentColor;opacity:.35;padding:0 .5em 0 0}
.CodeRay span.line-numbers{display:inline-block;margin-right:.75em}
.CodeRay .line-numbers strong{color:#000}
table.CodeRay{border-collapse:separate;border:0;margin-bottom:0;background:none}
table.CodeRay td{vertical-align:top;line-height:inherit}
table.CodeRay td.line-numbers{text-align:right}
table.CodeRay td.code{padding:0 0 0 .75em}
.CodeRay .debug{color:#fff !important;background:#000080 !important}
.CodeRay .annotation{color:#007}
.CodeRay .attribute-name{color:#000080}
.CodeRay .attribute-value{color:#700}
.CodeRay .binary{color:#509}
.CodeRay .comment{color:#998;font-style:italic}
.CodeRay .char{color:#04d}
.CodeRay .char .content{color:#04d}
.CodeRay .char .delimiter{color:#039}
.CodeRay .class{color:#458;font-weight:bold}
.CodeRay .complex{color:#a08}
.CodeRay .constant,.CodeRay .predefined-constant{color:#008080}
.CodeRay .color{color:#099}
.CodeRay .class-variable{color:#369}
.CodeRay .decorator{color:#b0b}
.CodeRay .definition{color:#099}
.CodeRay .delimiter{color:#000}
.CodeRay .doc{color:#970}
.CodeRay .doctype{color:#34b}
.CodeRay .doc-string{color:#d42}
.CodeRay .escape{color:#666}
.CodeRay .entity{color:#800}
.CodeRay .error{color:#808}
.CodeRay .exception{color:inherit}
.CodeRay .filename{color:#099}
.CodeRay .function{color:#900;font-weight:bold}
.CodeRay .global-variable{color:#008080}
.CodeRay .hex{color:#058}
.CodeRay .integer,.CodeRay .float{color:#099}
.CodeRay .include{color:#555}
.CodeRay .inline{color:#000}
.CodeRay .inline .inline{background:#ccc}
.CodeRay .inline .inline .inline{background:#bbb}
.CodeRay .inline .inline-delimiter{color:#d14}
.CodeRay .inline-delimiter{color:#d14}
.CodeRay .important{color:#555;font-weight:bold}
.CodeRay .interpreted{color:#b2b}
.CodeRay .instance-variable{color:#008080}
.CodeRay .label{color:#970}
.CodeRay .local-variable{color:#963}
.CodeRay .octal{color:#40e}
.CodeRay .predefined{color:#369}
.CodeRay .preprocessor{color:#579}
.CodeRay .pseudo-class{color:#555}
.CodeRay .directive{font-weight:bold}
.CodeRay .type{font-weight:bold}
.CodeRay .predefined-type{color:inherit}
.CodeRay .reserved,.CodeRay .keyword {color:#000;font-weight:bold}
.CodeRay .key{color:#808}
.CodeRay .key .delimiter{color:#606}
.CodeRay .key .char{color:#80f}
.CodeRay .value{color:#088}
.CodeRay .regexp .delimiter{color:#808}
.CodeRay .regexp .content{color:#808}
.CodeRay .regexp .modifier{color:#808}
.CodeRay .regexp .char{color:#d14}
.CodeRay .regexp .function{color:#404;font-weight:bold}
.CodeRay .string{color:#d20}
.CodeRay .string .string .string{background:#ffd0d0}
.CodeRay .string .content{color:#d14}
.CodeRay .string .char{color:#d14}
.CodeRay .string .delimiter{color:#d14}
.CodeRay .shell{color:#d14}
.CodeRay .shell .delimiter{color:#d14}
.CodeRay .symbol{color:#990073}
.CodeRay .symbol .content{color:#a60}
.CodeRay .symbol .delimiter{color:#630}
.CodeRay .tag{color:#008080}
.CodeRay .tag-special{color:#d70}
.CodeRay .variable{color:#036}
.CodeRay .insert{background:#afa}
.CodeRay .delete{background:#faa}
.CodeRay .change{color:#aaf;background:#007}
.CodeRay .head{color:#f8f;background:#505}
.CodeRay .insert .insert{color:#080}
.CodeRay .delete .delete{color:#800}
.CodeRay .change .change{color:#66f}
.CodeRay .head .head{color:#f4f}
</style>
</head>
<body class="article toc2 toc-left">
<div id="header">
<h1>TransPi Manual</h1>
<div class="details">
<span id="author" class="author">Ramón Rivera-Vicéns</span><br>
<span id="revnumber">version 1.0.0,</span>
<span id="revdate">2021-02-16</span>
</div>
<div id="toc" class="toc2">
<div id="toctitle">TransPi Manual</div>
<ul class="sectlevel1">
<li><a href="#_description">1. Description</a>
<ul class="sectlevel2">
<li><a href="#_programs_used">1.1. Programs used</a></li>
</ul>
</li>
<li><a href="#_installing_transpi">2. Installing TransPi</a>
<ul class="sectlevel2">
<li><a href="#_requirements">2.1. Requirements</a></li>
<li><a href="#_downloading_transpi">2.2. Downloading TransPi</a></li>
<li><a href="#_configuration">2.3. Configuration</a></li>
</ul>
</li>
<li><a href="#_running_transpi">3. Running TransPi</a>
<ul class="sectlevel2">
<li><a href="#_full_analysis_all">3.1. Full analysis (<code>--all</code>)</a></li>
<li><a href="#_other_options">3.2. Other options</a>
<ul class="sectlevel3">
<li><a href="#_onlyasm">3.2.1. <code>--onlyAsm</code></a></li>
<li><a href="#_onlyevi">3.2.2. <code>--onlyEvi</code></a></li>
<li><a href="#_onlyann">3.2.3. <code>--onlyAnn</code></a></li>
</ul>
</li>
<li><a href="#_using_profiles">3.3. Using <code>-profiles</code></a></li>
</ul>
</li>
<li><a href="#_results">4. Results</a>
<ul class="sectlevel2">
<li><a href="#_directories">4.1. Directories</a>
<ul class="sectlevel3">
<li><a href="#_results_2">4.1.1. <code>results</code></a></li>
<li><a href="#_work">4.1.2. <code>work</code></a></li>
</ul>
</li>
<li><a href="#_figures">4.2. Figures</a></li>
<li><a href="#_report">4.3. Report</a></li>
</ul>
</li>
<li><a href="#_additional_options">5. Additional options</a>
<ul class="sectlevel2">
<li><a href="#_output_options">5.1. Output options</a></li>
<li><a href="#_additional_analyses">5.2. Additional analyses</a></li>
<li><a href="#_skip_options">5.3. Skip options</a></li>
<li><a href="#_other_parameters">5.4. Other parameters</a></li>
</ul>
</li>
<li><a href="#_examples">6. Examples</a>
<ul class="sectlevel2">
<li><a href="#_profiles">6.1. Profiles</a>
<ul class="sectlevel3">
<li><a href="#_local_conda">6.1.1. Local conda</a></li>
<li><a href="#_individual_conda">6.1.2. Individual conda</a></li>
<li><a href="#_containers">6.1.3. Containers</a></li>
</ul>
</li>
<li><a href="#_other_examples">6.2. Other examples</a>
<ul class="sectlevel3">
<li><a href="#_filtering">6.2.1. Filtering</a></li>
<li><a href="#_busco_distribution">6.2.2. BUSCO distribution</a></li>
<li><a href="#_onlyevi_2">6.2.3. <code>--onlyEvi</code></a></li>
<li><a href="#_onlyann_2">6.2.4. <code>--onlyAnn</code></a></li>
<li><a href="#_skip_options_2">6.2.5. Skip options</a></li>
<li><a href="#_extra_annotation_steps">6.2.6. Extra annotation steps</a></li>
<li><a href="#_full_run_and_extra_annotation">6.2.7. Full run and extra annotation</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#_extra_information">7. Extra information</a>
<ul class="sectlevel2">
<li><a href="#_resume">7.1. <code>-resume</code></a></li>
<li><a href="#_template_nextflow_config">7.2. <code>template.nextflow.config</code></a>
<ul class="sectlevel3">
<li><a href="#_resources">7.2.1. Resources</a></li>
<li><a href="#_data">7.2.2. Data</a></li>
</ul>
</li>
<li><a href="#_custom_profiles">7.3. Custom profiles</a></li>
<li><a href="#_local_nextflow">7.4. Local nextflow</a></li>
<li><a href="#_real_time_monitoring">7.5. Real Time Monitoring</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div id="content">
<div id="preamble">
<div class="sectionbody">
<div class="paragraph">
<p>How to use TransPi</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_description"><a class="link" href="#_description">1. Description</a></h2>
<div class="sectionbody">
<div class="paragraph">
<p><strong>TransPi – a comprehensive TRanscriptome ANalysiS PIpeline for de novo transcriptome assembly</strong></p>
</div>
<div class="paragraph">
<p>TransPi provides a useful resource for the generation of de novo transcriptome assemblies,
with minimum user input but without losing the ability of a thorough analysis.</p>
</div>
<div class="ulist">
<ul>
<li>
<p>For more info see the <a href="https://doi.org/10.1101/2021.02.18.431773">Preprint</a></p>
</li>
<li>
<p>Code available at <a href="https://www.github.com/palmuc/TransPi">GitHub</a></p>
</li>
<li>
<p>Author: Ramón Rivera-Vicéns</p>
<div class="ulist">
<ul>
<li>
<p><a href="https://twitter.com/rerv787">Twitter</a></p>
</li>
</ul>
</div>
</li>
</ul>
</div>
<div class="sect2">
<h3 id="_programs_used"><a class="link" href="#_programs_used">1.1. Programs used</a></h3>
<div class="ulist">
<ul>
<li>
<p>List of programs use by TransPi:</p>
<div class="ulist">
<ul>
<li>
<p>FastQC</p>
</li>
<li>
<p>fastp</p>
</li>
<li>
<p>sortmerna</p>
</li>
<li>
<p>rnaSPADES</p>
</li>
<li>
<p>SOAP</p>
</li>
<li>
<p>Trinity</p>
</li>
<li>
<p>Velvet</p>
</li>
<li>
<p>Oases</p>
</li>
<li>
<p>TransABySS</p>
</li>
<li>
<p>rnaQUAST</p>
</li>
<li>
<p>EvidentialGene</p>
</li>
<li>
<p>CD-Hit</p>
</li>
<li>
<p>Exonerate</p>
</li>
<li>
<p>Blast</p>
</li>
<li>
<p>BUSCO</p>
</li>
<li>
<p>Psytrans</p>
</li>
<li>
<p>Trandecoder</p>
</li>
<li>
<p>Trinotate</p>
</li>
<li>
<p>Diamond</p>
</li>
<li>
<p>Hmmer</p>
</li>
<li>
<p>Bowtie2</p>
</li>
<li>
<p>rnammer</p>
</li>
<li>
<p>tmhmm</p>
</li>
<li>
<p>signalP</p>
</li>
<li>
<p>iPath</p>
</li>
<li>
<p>SQLite</p>
</li>
<li>
<p>R</p>
</li>
<li>
<p>Python</p>
</li>
</ul>
</div>
</li>
<li>
<p>Databases used by TransPi:</p>
<div class="ulist">
<ul>
<li>
<p>Swissprot</p>
</li>
<li>
<p>Uniprot custom database (e.g. all metazoan proteins)</p>
</li>
<li>
<p>Pfam</p>
</li>
</ul>
</div>
</li>
</ul>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_installing_transpi"><a class="link" href="#_installing_transpi">2. Installing TransPi</a></h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_requirements"><a class="link" href="#_requirements">2.1. Requirements</a></h3>
<div class="ulist">
<ul>
<li>
<p>System: Linux OS</p>
</li>
<li>
<p>Data type: Paired-end reads</p>
<div class="literalblock">
<div class="content">
<pre>Example:
    IndA_R1.fastq.gz, IndA_R2.fastq.gz</pre>
</div>
</div>
</li>
</ul>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
Make sure reads end with <code>_R1.fastq.gz</code> and <code>_R2.fastq.gz</code>.
Multiple individuals can be run at the same time.
</td>
</tr>
</table>
</div>
</div>
<div class="sect2">
<h3 id="_downloading_transpi"><a class="link" href="#_downloading_transpi">2.2. Downloading TransPi</a></h3>
<div class="paragraph">
<p>1- Clone the repository</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">git clone https://github.com/palmuc/TransPi.git</code></pre>
</div>
</div>
<div class="paragraph">
<p>2- Move to the TransPi directory</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">cd TransPi</code></pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_configuration"><a class="link" href="#_configuration">2.3. Configuration</a></h3>
<div class="paragraph">
<p>TransPi requires various databases to run. The precheck script will installed the databases and software, if necessary, to run the tool.
The precheck run needs a <code>PATH</code> as an argument for installing (locally) all the databases the pipeline needs.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code>bash precheck_TransPi.sh /YOUR/PATH/HERE/</code></pre>
</div>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
This process may take a while depending on the options you select. Step that takes longer is downloading, if desired, the entire metazoan proteins from UniProt (6Gb).
Other processes and databases are relatively fast depending on internet connection.
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>Once the precheck run is done it will create a file named <code>nextflow.config</code> that contains the various <code>PATH</code> for the databases.
If  selected, it will also have the local conda environment <code>PATH</code>.</p>
</div>
<div class="paragraph">
<p>The <code>nextflow.config</code> file has also other important parameters for pipeline execution that will be discussed further
in the following sections.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_running_transpi"><a class="link" href="#_running_transpi">3. Running TransPi</a></h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_full_analysis_all"><a class="link" href="#_full_analysis_all">3.1. Full analysis (<code>--all</code>)</a></h3>
<div class="paragraph">
<p>After the successful run of the precheck script you are set to run TransPi.</p>
</div>
<div class="paragraph">
<p>We recommend to run TransPi with the option <code>--all</code> where it will do the complete analysis, from raw reads filtering to annotation.
Other options described below.</p>
</div>
<div class="paragraph">
<p>To run the complete pipeline.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --reads '/YOUR/READS/PATH/HERE/*_R[1,2].fastq.gz' \
     --k 25,41,53 --maxReadLen 75 -profile conda --myConda</code></pre>
</div>
</div>
<div class="paragraph">
<p>Arguments explanations:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="text">--all                Run full TransPi analysis
--reads              PATH to the paired-end reads
--k                  kmers list to use for the assemblies
--maxReadLen         Max read length in the library
-profile             Use conda to run the analyses
--myConda            Use local conda installed by precheck run</code></pre>
</div>
</div>
<div class="admonitionblock caution">
<table>
<tr>
<td class="icon">
<i class="fa icon-caution" title="Caution"></i>
</td>
<td class="content">
<div class="paragraph">
<p>If you combined multiple libraries of the same individual to create a reference transcriptome, which will be later use in downstream analyses (e.g. Differential Expression),
make sure the kmer list is based on the length for the shortest read library and the <code>maxReadLen</code> based on the longest read length.</p>
</div>
<div class="paragraph">
<p>Example: Combining reads of 100bp with 125bp</p>
</div>
<div class="sidebarblock">
<div class="content">
<div class="paragraph">
<p>--k 25,41,53,61 --maxReadLen 125</p>
</div>
</div>
</div>
</td>
</tr>
</table>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
<div class="paragraph">
<p>You can run multiple samples at the same time</p>
</div>
</td>
</tr>
</table>
</div>
</div>
<div class="sect2">
<h3 id="_other_options"><a class="link" href="#_other_options">3.2. Other options</a></h3>
<div class="sect3">
<h4 id="_onlyasm"><a class="link" href="#_onlyasm">3.2.1. <code>--onlyAsm</code></a></h4>
<div class="paragraph">
<p>Run only the Assemblies and EvidentialGene analysis.</p>
</div>
<div class="paragraph">
<p>Example for <code>--onlyAsm</code>:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --onlyAsm --reads '/home/rrivera/TransPi/reads/*_R[1,2].fastq.gz' \
     --k 25,41,53 --maxReadLen 75 -profile conda --myConda</code></pre>
</div>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
You can run multiple samples at the same time
</td>
</tr>
</table>
</div>
</div>
<div class="sect3">
<h4 id="_onlyevi"><a class="link" href="#_onlyevi">3.2.2. <code>--onlyEvi</code></a></h4>
<div class="paragraph">
<p>Run only the Evidential Gene analysis</p>
</div>
<div class="paragraph">
<p>Example for <code>--onlyEvi</code>:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --onlyEvi -profile conda --myConda</code></pre>
</div>
</div>
<div class="admonitionblock important">
<table>
<tr>
<td class="icon">
<i class="fa icon-important" title="Important"></i>
</td>
<td class="content">
TransPi looks for a directory named <code>onlyEvi</code>. It expects one file per sample to perform the reduction. The file should have all the assemblies concatenated into one.
</td>
</tr>
</table>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
You can run multiple samples at the same time
</td>
</tr>
</table>
</div>
</div>
<div class="sect3">
<h4 id="_onlyann"><a class="link" href="#_onlyann">3.2.3. <code>--onlyAnn</code></a></h4>
<div class="paragraph">
<p>Run only the Annotation analysis (starting from a final assembly)</p>
</div>
<div class="paragraph">
<p>Example for <code>--onlyAnn</code>:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --onlyAnn -profile conda --myConda</code></pre>
</div>
</div>
<div class="admonitionblock important">
<table>
<tr>
<td class="icon">
<i class="fa icon-important" title="Important"></i>
</td>
<td class="content">
TransPi looks for a directory named <code>onlyAnn</code>. It expects one file per sample to perform the annotation.
</td>
</tr>
</table>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
You can run multiple samples at the same time
</td>
</tr>
</table>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_using_profiles"><a class="link" href="#_using_profiles">3.3. Using <code>-profiles</code></a></h3>
<div class="paragraph">
<p>TransPi can also use docker, singularity, and individual conda installations (i.e. per process) to deploy the pipeline.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="text">test                Run TransPi with a test dataset
conda               Run TransPi with conda.
docker              Run TransPi with docker container
singularity         Run TransPi with singularity container with all the necessary tools
TransPiContainer    Run TransPi with a single container with all tools</code></pre>
</div>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
<div class="paragraph">
<p>Multiple profiles can be specified (comma separated)</p>
</div>
<div class="sidebarblock">
<div class="content">
<div class="paragraph">
<p>Example: <code>-profile test,singularity</code></p>
</div>
</div>
</div>
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>Refer to <strong>Section 6</strong> of this manual for further details on deployment of TransPi using other profiles.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_results"><a class="link" href="#_results">4. Results</a></h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_directories"><a class="link" href="#_directories">4.1. Directories</a></h3>
<div class="sect3">
<h4 id="_results_2"><a class="link" href="#_results_2">4.1.1. <code>results</code></a></h4>
<div class="paragraph">
<p>After a successful run of TransPi the results are saved in a directory called <code>results</code>. This directory is divided into multiple directories for each major step of the pipeline.</p>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
Directories will be created based on the options selected in the pipeline execution
</td>
</tr>
</table>
</div>
<div class="hdlist">
<table>
<tr>
<td class="hdlist1">
fastqc
</td>
<td class="hdlist2">
<p>Fastqc html files</p>
</td>
</tr>
<tr>
<td class="hdlist1">
filter
</td>
<td class="hdlist2">
<p>Filter step html files</p>
</td>
</tr>
<tr>
<td class="hdlist1">
rRNA_reads
</td>
<td class="hdlist2">
<p>Info and reads of rRNA removal process</p>
</td>
</tr>
<tr>
<td class="hdlist1">
normalization
</td>
<td class="hdlist2">
<p>Normalized reads files</p>
</td>
</tr>
<tr>
<td class="hdlist1">
saveReads
</td>
<td class="hdlist2">
<p>Folder with reads saved from the filter and normalization processes</p>
</td>
</tr>
<tr>
<td class="hdlist1">
assemblies
</td>
<td class="hdlist2">
<p>All individual assemblies</p>
</td>
</tr>
<tr>
<td class="hdlist1">
evigene
</td>
<td class="hdlist2">
<p>Non-redundant final transcriptome (ends with name <code>.combined.okay.fa</code>)</p>
</td>
</tr>
<tr>
<td class="hdlist1">
rnaQuast
</td>
<td class="hdlist2">
<p>rnaQUAST output</p>
</td>
</tr>
<tr>
<td class="hdlist1">
mapping
</td>
<td class="hdlist2">
<p>Mapping results</p>
</td>
</tr>
<tr>
<td class="hdlist1">
busco3
</td>
<td class="hdlist2">
<p>BUSCO V3 results</p>
</td>
</tr>
<tr>
<td class="hdlist1">
busco4
</td>
<td class="hdlist2">
<p>BUSCO V4 results</p>
</td>
</tr>
<tr>
<td class="hdlist1">
transdecoder
</td>
<td class="hdlist2">
<p>Transdecoder results</p>
</td>
</tr>
<tr>
<td class="hdlist1">
trinotate
</td>
<td class="hdlist2">
<p>Annotation results</p>
</td>
</tr>
<tr>
<td class="hdlist1">
report
</td>
<td class="hdlist2">
<p>Interactive report of TransPi</p>
</td>
</tr>
<tr>
<td class="hdlist1">
figures
</td>
<td class="hdlist2">
<p>Figures created by TransPi (BUSCO comparison, Annotation, GO, etc)</p>
</td>
</tr>
<tr>
<td class="hdlist1">
stats
</td>
<td class="hdlist2">
<p>Basic stats from all steps of TransPi</p>
</td>
</tr>
<tr>
<td class="hdlist1">
pipeline_info
</td>
<td class="hdlist2">
<p>Nextflow report, trace file and others</p>
</td>
</tr>
<tr>
<td class="hdlist1">
versions.txt
</td>
<td class="hdlist2">
<p>File with all versionts of the tools used by TransPi</p>
</td>
</tr>
</table>
</div>
<div class="sidebarblock">
<div class="content">
<div class="title">NOTES</div>
<div class="ulist">
<ul>
<li>
<p>Name of output directory can be changed by using the <code>--outdir</code> parameter when executing the pipeline. Example <code>--outdir Results_SampleA</code>.</p>
</li>
<li>
<p>If multiple samples are run, each directory will have all files together but each one with a unique sample name.</p>
</li>
</ul>
</div>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_work"><a class="link" href="#_work">4.1.2. <code>work</code></a></h4>
<div class="paragraph">
<p>A directory called <code>work</code> is also created when running TransPi. It contains all the Nextflow working files, TransPi results and intermediate files.</p>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
Directory <code>work</code> can be removed after the pipeline is done since all important files are stored in the <code>results</code> directory.
</td>
</tr>
</table>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_figures"><a class="link" href="#_figures">4.2. Figures</a></h3>
<div class="paragraph">
<p>TransPi produces multiple figures that are stored in the results directory.</p>
</div>
<div class="paragraph">
<p>Example:</p>
</div>
<div class="paragraph">
<p><span class="image center Uniprot"><img src="https://sync.palmuc.org/index.php/s/kxetdGiNiSyHzrg/preview" alt="UniProt" width="800" height="400"></span></p>
</div>
</div>
<div class="sect2">
<h3 id="_report"><a class="link" href="#_report">4.3. Report</a></h3>
<div class="paragraph">
<p>TransPi creates an interactive custom HTML report for ease data exploration.</p>
</div>
<div class="paragraph">
<p>Report <a href="https://sync.palmuc.org/index.php/s/XCxeCNwAfParBHX">Sponge transcriptome</a></p>
</div>
<div class="sidebarblock">
<div class="content">
<div class="title">NOTE</div>
<div class="ulist">
<ul>
<li>
<p>Example report here is a PDF file and not a HTML file. However the original HTML file with interactive visualization (i.e. as generated in TransPi) can be downloaded <a href="https://sync.palmuc.org/index.php/s/nP3TKPawmoX4xqL">here</a></p>
</li>
</ul>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_additional_options"><a class="link" href="#_additional_options">5. Additional options</a></h2>
<div class="sectionbody">
<div class="paragraph">
<p>There are other parameters that can be changed when executing TransPi.</p>
</div>
<div class="sect2">
<h3 id="_output_options"><a class="link" href="#_output_options">5.1. Output options</a></h3>
<div class="hdlist">
<table>
<tr>
<td class="hdlist1">
<code>--outdir</code>
</td>
<td class="hdlist2">
<p>name of output directory. Example: <code>--outdir Sponges_150</code>.
Default "results"</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>-w, -work</code>
</td>
<td class="hdlist2">
<p>name of working directory. Example: <code>-work Sponges_work</code>. Only one dash is needed for <code>-work</code> since it is a nextflow function.</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--tracedir</code>
</td>
<td class="hdlist2">
<p>Name for directory to save pipeline trace files.
Default "pipeline_info"</p>
</td>
</tr>
</table>
</div>
</div>
<div class="sect2">
<h3 id="_additional_analyses"><a class="link" href="#_additional_analyses">5.2. Additional analyses</a></h3>
<div class="hdlist">
<table>
<tr>
<td class="hdlist1">
<code>--rRNAfilter</code>
</td>
<td class="hdlist2">
<p>Remove rRNA from sequences. Requires option --rRNAdb</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--rRNAdb</code>
</td>
<td class="hdlist2">
<p>PATH to database of rRNA sequences to use for filtering of rRNA. Default ""</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--filterSpecies</code>
</td>
<td class="hdlist2">
<p>Perform psytrans filtering of transcriptome. Requires options <code>--host</code> and <code>--symbiont</code></p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--host</code>
</td>
<td class="hdlist2">
<p>Host (or similar) protein file.</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--symbiont</code>
</td>
<td class="hdlist2">
<p>Symbionts (or similar) protein files</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--psyval</code>
</td>
<td class="hdlist2">
<p>Psytrans value to train model. Default "160"</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--allBuscos</code>
</td>
<td class="hdlist2">
<p>Run BUSCO analysis in all assemblies</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--rescueBusco</code>
</td>
<td class="hdlist2">
<p>Generate BUSCO distribution analysis</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--minPerc</code>
</td>
<td class="hdlist2">
<p>Mininmum percentage of assemblers require for the BUSCO distribution.
Default ".70"</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--shortTransdecoder</code>
</td>
<td class="hdlist2">
<p>Run Transdecoder without the homology searches</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--withSignalP</code>
</td>
<td class="hdlist2">
<p>Include SignalP for the annotation. Needs manual installation of CBS-DTU tools.
Default "false"</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--rnam</code>
</td>
<td class="hdlist2">
<p>PATH to Rnammer software. Default ""</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--withTMHMM</code>
</td>
<td class="hdlist2">
<p>Include TMHMM for the annotation. Needs manual installation of CBS-DTU tools.
Default "false"</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--tmhmm</code>
</td>
<td class="hdlist2">
<p>PATH to TMHMM software. Default ""</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--withRnammer</code>
</td>
<td class="hdlist2">
<p>Include Rnammer for the annotation. Needs manual installation of CBS-DTU tools.
Default "false"</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--rnam</code>
</td>
<td class="hdlist2">
<p>PATH to Rnammer software. Default ""</p>
</td>
</tr>
</table>
</div>
</div>
<div class="sect2">
<h3 id="_skip_options"><a class="link" href="#_skip_options">5.3. Skip options</a></h3>
<div class="hdlist">
<table>
<tr>
<td class="hdlist1">
<code>--skipEvi</code>
</td>
<td class="hdlist2">
<p>Skip EvidentialGene run in --onlyAsm option. Default "false"</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--skipQC</code>
</td>
<td class="hdlist2">
<p>Skip FastQC step. Default "false"</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--skipFilter</code>
</td>
<td class="hdlist2">
<p>Skip fastp filtering step. Default "false"</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--skipKegg</code>
</td>
<td class="hdlist2">
<p>Skip kegg analysis. Default "false"</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--skipReport</code>
</td>
<td class="hdlist2">
<p>Skip generation of final TransPi report. Default "false"</p>
</td>
</tr>
</table>
</div>
</div>
<div class="sect2">
<h3 id="_other_parameters"><a class="link" href="#_other_parameters">5.4. Other parameters</a></h3>
<div class="hdlist">
<table>
<tr>
<td class="hdlist1">
<code>--minQual</code>
</td>
<td class="hdlist2">
<p>Minimum quality score for fastp filtering. Default "25"</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--pipeInstall</code>
</td>
<td class="hdlist2">
<p>PATH to TransPi directory. Default "". If precheck is used this will be added to the nextflow.config automatically.</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--myCondaInstall</code>
</td>
<td class="hdlist2">
<p>PATH to local conda environment of TransPi. Default "". Requires use of --myConda.</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--myConda</code>
</td>
<td class="hdlist2">
<p>Make TransPi use a local conda environemt with all the tools (generated with precheck)</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--envCacheDir</code>
</td>
<td class="hdlist2">
<p>PATH for environment cache directory (either conda or containers). Default "Launch directory of pipeline"</p>
</td>
</tr>
<tr>
<td class="hdlist1">
<code>--getVersions</code>
</td>
<td class="hdlist2">
<p>Get software versions. Default "false"</p>
</td>
</tr>
</table>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_examples"><a class="link" href="#_examples">6. Examples</a></h2>
<div class="sectionbody">
<div class="paragraph">
<p>Here are some examples on how to deploy TransPi depending on the method to use (e.g. conda) and the analyses to be performed.</p>
</div>
<div class="sect2">
<h3 id="_profiles"><a class="link" href="#_profiles">6.1. Profiles</a></h3>
<div class="paragraph">
<p>You can use TransPi either with:
- a local conda environment (from precheck);
- individual conda environment per process;
- docker or singularity</p>
</div>
<div class="sect3">
<h4 id="_local_conda"><a class="link" href="#_local_conda">6.1.1. Local conda</a></h4>
<div class="paragraph">
<p>This way of executing TransPi assumes that you installed conda locally and created the TransPi conda environment.
All these is done automatically, if desired, with the precheck script.</p>
</div>
<div class="paragraph">
<p><strong>Example:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --maxReadLen 150 --k 25,35,55,75,85 \
    --reads '/YOUR/PATH/HERE/*_R[1,2].fastq.gz' --outdir Results_Acropora \
    -profile conda --myConda</code></pre>
</div>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
<code>myConda</code> tells TransPi that the conda environment to use is the local one. The PATH for this environment is taken
from the <code>nextflow.config</code>, specifically the parameter <code>myCondaInstall</code>.
</td>
</tr>
</table>
</div>
</div>
<div class="sect3">
<h4 id="_individual_conda"><a class="link" href="#_individual_conda">6.1.2. Individual conda</a></h4>
<div class="paragraph">
<p>You also have the option to run TransPi with individual conda environments for each process. To use this you only specify the profile for conda.</p>
</div>
<div class="paragraph">
<p><strong>Example:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --maxReadLen 150 --k 25,35,55,75,85 \
    --reads '/YOUR/PATH/HERE/*_R[1,2].fastq.gz' --outdir Results_Acropora \
    -profile conda</code></pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_containers"><a class="link" href="#_containers">6.1.3. Containers</a></h4>
<div class="paragraph">
<p>Docker or singularity can also be use for deploying TransPi. You can either use individual containers for each process or a TransPi container with all the tools.</p>
</div>
<div class="sect4">
<h5 id="_individual"><a class="link" href="#_individual">Individual</a></h5>
<div class="paragraph">
<p>To use individual containers:</p>
</div>
<div class="paragraph">
<p><strong>Example for docker:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --maxReadLen 150 --k 25,35,55,75,85 \
    --reads '/YOUR/PATH/HERE/*_R[1,2].fastq.gz' --outdir Results_Acropora \
    -profile docker</code></pre>
</div>
</div>
<div class="paragraph">
<p><strong>Example for singularity:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --maxReadLen 150 --k 25,35,55,75,85 \
    --reads '/YOUR/PATH/HERE/*_R[1,2].fastq.gz' --outdir Results_Acropora \
    -profile singularity</code></pre>
</div>
</div>
</div>
<div class="sect4">
<h5 id="_transpi_container"><a class="link" href="#_transpi_container">TransPi container</a></h5>
<div class="paragraph">
<p>To use the TransPi container with all the tools you need to use the profile <code>TransPiContainer</code>.</p>
</div>
<div class="paragraph">
<p><strong>Example for docker:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --maxReadLen 150 --k 25,35,55,75,85 \
    --reads '/YOUR/PATH/HERE/*_R[1,2].fastq.gz' --outdir Results_Acropora \
    -profile docker,TransPiContainer</code></pre>
</div>
</div>
<div class="paragraph">
<p><strong>Example for singularity:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --maxReadLen 150 --k 25,35,55,75,85 \
    --reads '/YOUR/PATH/HERE/*_R[1,2].fastq.gz' --outdir Results_Acropora \
    -profile singularity,TransPiContainer</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_other_examples"><a class="link" href="#_other_examples">6.2. Other examples</a></h3>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
Order of commands is not important.
</td>
</tr>
</table>
</div>
<div class="sect3">
<h4 id="_filtering"><a class="link" href="#_filtering">6.2.1. Filtering</a></h4>
<div class="paragraph">
<p><strong>Scenario:</strong></p>
</div>
<div class="hdlist">
<table>
<tr>
<td class="hdlist1">
Sample
</td>
<td class="hdlist2">
<p>Coral sample</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Read length
</td>
<td class="hdlist2">
<p>150bp</p>
</td>
</tr>
<tr>
<td class="hdlist1">
TransPi mode
</td>
<td class="hdlist2">
<p>--all</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Kmers
</td>
<td class="hdlist2">
<p>25,35,55,75,85</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Reads PATH
</td>
<td class="hdlist2">
<p>/YOUR/PATH/HERE/*_R[1,2].fastq.gz</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Output directory
</td>
<td class="hdlist2">
<p>Results_Acropora</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Work directory
</td>
<td class="hdlist2">
<p>work_acropora</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Engine
</td>
<td class="hdlist2">
<p>local conda</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Filter species
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
host
</td>
<td class="hdlist2">
<p>scleractinian proteins</p>
</td>
</tr>
<tr>
<td class="hdlist1">
symbiont
</td>
<td class="hdlist2">
<p>symbiodinium proteins</p>
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><strong>Command:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --maxReadLen 150 --k 25,35,55,75,85 \
    --reads '/YOUR/PATH/HERE/*_R[1,2].fastq.gz' --outdir Results_Acropora \
    -w work_acropora -profile condas --myConda --filterSpecies \
    --host /YOUR/PATH/HERE/uniprot-Scleractinia.fasta \
    --symbiont /YOUR/PATH/HERE/uniprot-symbiodinium.fasta</code></pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_busco_distribution"><a class="link" href="#_busco_distribution">6.2.2. BUSCO distribution</a></h4>
<div class="paragraph">
<p><strong>Scenario:</strong></p>
</div>
<div class="hdlist">
<table>
<tr>
<td class="hdlist1">
Sample
</td>
<td class="hdlist2">
<p>SampleA</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Read length
</td>
<td class="hdlist2">
<p>100bp</p>
</td>
</tr>
<tr>
<td class="hdlist1">
TransPi mode
</td>
<td class="hdlist2">
<p>--all</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Kmers
</td>
<td class="hdlist2">
<p>25,41,57,67</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Reads PATH
</td>
<td class="hdlist2">
<p>/YOUR/PATH/HERE/SampleA/*_R[1,2].fastq.gz</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Output directory
</td>
<td class="hdlist2">
<p>Results_SampleA</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Engine
</td>
<td class="hdlist2">
<p>conda</p>
</td>
</tr>
<tr>
<td class="hdlist1">
All BUSCOs
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
BUSCO distribution
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><strong>Command:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --maxReadLen 100 --k 25,35,55,75,85 \
    --outdir Results_SampleA --reads '/YOUR/PATH/HERE/SampleA/*_R[1,2].fastq.gz' \
    -profile conda --allBuscos --buscoDist</code></pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_onlyevi_2"><a class="link" href="#_onlyevi_2">6.2.3. <code>--onlyEvi</code></a></h4>
<div class="paragraph">
<p><strong>Scenario:</strong></p>
</div>
<div class="hdlist">
<table>
<tr>
<td class="hdlist1">
Sample
</td>
<td class="hdlist2">
<p>Assemblies from multiple assemblers and kmers</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Read length
</td>
<td class="hdlist2">
<p>50bp</p>
</td>
</tr>
<tr>
<td class="hdlist1">
TransPi mode
</td>
<td class="hdlist2">
<p>--onlyEvi</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Kmers
</td>
<td class="hdlist2">
<p>25,33,37</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Reads PATH
</td>
<td class="hdlist2">
<p>/YOUR/PATH/HERE/*_R[1,2].fastq.gz</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Output directory
</td>
<td class="hdlist2">
<p>Reduction_results</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Engine
</td>
<td class="hdlist2">
<p>conda</p>
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><strong>Command:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --onlyEvi --outdir Reduction_results \
    -profile conda</code></pre>
</div>
</div>
<div class="sidebarblock">
<div class="content">
<div class="title">NOTES</div>
<div class="ulist">
<ul>
<li>
<p>A directory named <code>onlyEvi</code> is needed for this option with the transcriptome to perform the reduction.</p>
</li>
</ul>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
You can do multiple transcriptomes at the same time. Each file should have a unique name.
</td>
</tr>
</table>
</div>
<div class="ulist">
<ul>
<li>
<p>No need to specify reads PATH, length, cutoff, and kmers when using the <code>--onlyEvi</code>.</p>
</li>
</ul>
</div>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_onlyann_2"><a class="link" href="#_onlyann_2">6.2.4. <code>--onlyAnn</code></a></h4>
<div class="paragraph">
<p><strong>Scenario:</strong></p>
</div>
<div class="hdlist">
<table>
<tr>
<td class="hdlist1">
Sample
</td>
<td class="hdlist2">
<p>Transcriptome missing annotation</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Read length
</td>
<td class="hdlist2">
<p>100bp</p>
</td>
</tr>
<tr>
<td class="hdlist1">
TransPi mode
</td>
<td class="hdlist2">
<p>--onlyEvi</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Kmers
</td>
<td class="hdlist2">
<p>25,41,57,67</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Reads PATH
</td>
<td class="hdlist2">
<p>/YOUR/PATH/HERE/*_R[1,2].fastq.gz</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Output directory
</td>
<td class="hdlist2">
<p>Annotation_results</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Engine
</td>
<td class="hdlist2">
<p>singularity</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Container
</td>
<td class="hdlist2">
<p>TransPi container</p>
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><strong>Command:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --onlyAnn --outdir Annotation_results \
    -profile singularity,TransPiContainer</code></pre>
</div>
</div>
<div class="sidebarblock">
<div class="content">
<div class="title">NOTES</div>
<div class="ulist">
<ul>
<li>
<p>A directory named <code>onlyAnn</code> is needed for this option with the transcriptome to annotate.</p>
</li>
</ul>
</div>
<div class="admonitionblock tip">
<table>
<tr>
<td class="icon">
<i class="fa icon-tip" title="Tip"></i>
</td>
<td class="content">
You can do multiple transcriptomes (i.e. samples) at the same time. Each file should have a unique name.
</td>
</tr>
</table>
</div>
<div class="ulist">
<ul>
<li>
<p>No need to specify reads PATH, length, cutoff, and kmers when using the <code>--onlyAnn</code>.</p>
</li>
</ul>
</div>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_skip_options_2"><a class="link" href="#_skip_options_2">6.2.5. Skip options</a></h4>
<div class="paragraph">
<p><strong>Scenario:</strong></p>
</div>
<div class="hdlist">
<table>
<tr>
<td class="hdlist1">
Sample
</td>
<td class="hdlist2">
<p>Coral sample</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Read length
</td>
<td class="hdlist2">
<p>150bp</p>
</td>
</tr>
<tr>
<td class="hdlist1">
TransPi mode
</td>
<td class="hdlist2">
<p>--all</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Kmers
</td>
<td class="hdlist2">
<p>25,35,55,75,85</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Reads PATH
</td>
<td class="hdlist2">
<p>/YOUR/PATH/HERE/*_R[1,2].fastq.gz</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Output directory
</td>
<td class="hdlist2">
<p>Results_Acropora</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Work directory
</td>
<td class="hdlist2">
<p>work_acropora</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Engine
</td>
<td class="hdlist2">
<p>docker</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Container
</td>
<td class="hdlist2">
<p>Individual containers</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Skip QC
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Skip Filter
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><strong>Command:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --maxReadLen 150 --k 25,35,55,75,85 \
    --reads '/YOUR/PATH/HERE/*_R[1,2].fastq.gz' --outdir Results_Acropora \
    -w work_acropora -profile docker \
    --skipQC --skipFilter</code></pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_extra_annotation_steps"><a class="link" href="#_extra_annotation_steps">6.2.6. Extra annotation steps</a></h4>
<div class="paragraph">
<p><strong>Scenario:</strong></p>
</div>
<div class="hdlist">
<table>
<tr>
<td class="hdlist1">
Sample
</td>
<td class="hdlist2">
<p>Mollusc sample</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Read length
</td>
<td class="hdlist2">
<p>150bp</p>
</td>
</tr>
<tr>
<td class="hdlist1">
TransPi mode
</td>
<td class="hdlist2">
<p>--all</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Kmers
</td>
<td class="hdlist2">
<p>25,35,55,75,85</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Reads PATH
</td>
<td class="hdlist2">
<p>/YOUR/PATH/HERE/*_R[1,2].fastq.gz</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Output directory
</td>
<td class="hdlist2">
<p>Results</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Engine
</td>
<td class="hdlist2">
<p>conda</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Skip QC
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
SignalP
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
TMHMM
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
RNAmmer
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><strong>Command:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --maxReadLen 150 --k 25,35,55,75,85 \
    --reads '/YOUR/PATH/HERE/*_R[1,2].fastq.gz' --outdir Results \
    -profile conda -myConda --skipQC --withSignalP --withTMHMM --withRnammer</code></pre>
</div>
</div>
<div class="sidebarblock">
<div class="content">
<div class="title">NOTE</div>
<div class="ulist">
<ul>
<li>
<p>This option requires manual installation of the CBS-DTU tools: signalP, tmhmm, and rnammer.</p>
</li>
<li>
<p>For more info visit <a href="https://services.healthtech.dtu.dk/software.php">CBS-DTU tools</a></p>
</li>
<li>
<p>It also assumes that the <code>PATH</code> for all the tools are in the <code>nextflow.config</code> file.</p>
</li>
</ul>
</div>
</div>
</div>
<div class="paragraph">
<p>If the <code>PATH</code> are not specify in the <code>nextflow.config</code> then you can either add them in the file or use the parameters from the command line.</p>
</div>
<div class="paragraph">
<p><strong>Example:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --maxReadLen 150 --k 25,35,55,75,85 \
    --reads '/YOUR/PATH/HERE/*_R[1,2].fastq.gz' --outdir Results \
    -profile conda --myConda --myCondaInstall &quot;/YOUR/PATH/HERE/TransPi&quot; --skipQC
    --withSignalP --signalp &quot;/YOUR/PATH/HERE/rnammer&quot;
    --withTMHMM --tmhmm &quot;/YOUR/PATH/HERE/tmhmm&quot;
    --withRnammer --rnam &quot;/YOUR/PATH/HERE/signalp&quot;</code></pre>
</div>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
In this example the local conda environment was not specified in the <code>nextflow.config</code> but as a command line argument.
</td>
</tr>
</table>
</div>
</div>
<div class="sect3">
<h4 id="_full_run_and_extra_annotation"><a class="link" href="#_full_run_and_extra_annotation">6.2.7. Full run and extra annotation</a></h4>
<div class="paragraph">
<p><strong>Scenario:</strong></p>
</div>
<div class="hdlist">
<table>
<tr>
<td class="hdlist1">
Sample
</td>
<td class="hdlist2">
<p>Coral sample</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Read length
</td>
<td class="hdlist2">
<p>150bp</p>
</td>
</tr>
<tr>
<td class="hdlist1">
TransPi mode
</td>
<td class="hdlist2">
<p>--all</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Kmers
</td>
<td class="hdlist2">
<p>25,35,55,75,85</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Reads PATH
</td>
<td class="hdlist2">
<p>/YOUR/PATH/HERE/*_R[1,2].fastq.gz</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Output directory
</td>
<td class="hdlist2">
<p>Results</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Engine
</td>
<td class="hdlist2">
<p>conda</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Skip QC
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
SignalP
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
TMHMM
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
RNAmmer
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Filter species
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
host
</td>
<td class="hdlist2">
<p>scleractinian proteins</p>
</td>
</tr>
<tr>
<td class="hdlist1">
symbiont
</td>
<td class="hdlist2">
<p>symbiodinium proteins</p>
</td>
</tr>
<tr>
<td class="hdlist1">
All BUSCOs
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
BUSCO distribution
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
Remove rRNA
</td>
<td class="hdlist2">
<p>on</p>
</td>
</tr>
<tr>
<td class="hdlist1">
rRNA database
</td>
<td class="hdlist2">
<p>/YOUR/PATH/HERE/silva_rRNA_file.fasta</p>
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p><strong>Command:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">nextflow run TransPi.nf --all --maxReadLen 150 --k 25,35,55,75,85 \
    --reads '/YOUR/PATH/HERE/*_R[1,2].fastq.gz' --outdir Results \
    -profile conda --skipQC --withSignalP --withTMHMM --withRnammer \
    --host /YOUR/PATH/HERE/uniprot-Scleractinia.fasta \
    --symbiont /YOUR/PATH/HERE/uniprot-symbiodinium.fasta
    --allBuscos --buscoDist --rRNAfilter \
    --rRNAdb &quot;/YOUR/PATH/HERE/silva_rRNA_file.fasta&quot;</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_extra_information"><a class="link" href="#_extra_information">7. Extra information</a></h2>
<div class="sectionbody">
<div class="paragraph">
<p>Here are some notes that can help in the execution of TransPi. Also some important considerations based Nextflow settings.
For more in detail information visit the <a href="https://www.nextflow.io/docs/latest/index.html">Nextflow documentation</a></p>
</div>
<div class="sect2">
<h3 id="_resume"><a class="link" href="#_resume">7.1. <code>-resume</code></a></h3>
<div class="paragraph">
<p>If an error occurs and you need to resume the pipeline just include the <code>-resume</code> option when calling the pipeline.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="CodeRay highlight"><code data-lang="bash">./nextflow run TransPi.nf --onlyAnn -profile conda -resume</code></pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_template_nextflow_config"><a class="link" href="#_template_nextflow_config">7.2. <code>template.nextflow.config</code></a></h3>
<div class="sect3">
<h4 id="_resources"><a class="link" href="#_resources">7.2.1. Resources</a></h4>
<div class="paragraph">
<p>The <code>template.nextflow.config</code> file has different configurations for the each program of the pipeline
(e.g. some with a lot of CPUs, others with a small amount of CPUs). You can modify this depending on the resources you have in your system.</p>
</div>
<div class="paragraph">
<p>Example:</p>
</div>
<div class="sidebarblock">
<div class="content">
<div class="paragraph">
<p>process {
    withLabel: big_cpus {
        cpus='30'
        memory='15 GB'
    }</p>
</div>
</div>
</div>
<div class="paragraph">
<p>In this case, all the processes using the label <code>big_cpus</code> will use 30 CPUs. If your system only has 20 please modify these values accordingly to avoid errors.</p>
</div>
<div class="admonitionblock note">
<table>
<tr>
<td class="icon">
<i class="fa icon-note" title="Note"></i>
</td>
<td class="content">
Setting the correct CPUs and RAM of your system is important because <code>nextflow</code> will start as many jobs as possible if the resources are available.
If you are in a VM with 120 CPUs, <code>nextfow</code> will be able to start four processes with the label <code>big_cpus</code>.
</td>
</tr>
</table>
</div>
</div>
<div class="sect3">
<h4 id="_data"><a class="link" href="#_data">7.2.2. Data</a></h4>
<div class="paragraph">
<p>The precheck is designed to create a new <code>nextflow.config</code> every time is run with with the <code>PATH</code> to the databases.
You can modify the values that do not need editing for your analysis on the <code>template.nextflow.config</code>. This way you avoid doing the same changes to the <code>nextflow.config</code> after the precheck run.</p>
</div>
<div class="paragraph">
<p>Example: Modify the <code>template.nextflow.config</code> with your cluster info to avoid repeating these in the future.</p>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_custom_profiles"><a class="link" href="#_custom_profiles">7.3. Custom profiles</a></h3>
<div class="paragraph">
<p>We are using <a href="https://slurm.schedmd.com/documentation.html">SLURM</a> as our workload manager in our server.
Thus we have custom profiles for the submission of jobs. For example our <code>nextflow.config</code> has the following lines in the <code>profiles</code> section.</p>
</div>
<div class="sidebarblock">
<div class="content">
<div class="paragraph">
<p>profiles {
    palmuc {
        process {
            executor='slurm'
            clusterOptions='--clusters=inter --partition=bigNode --qos=low'
        }
    }
}</p>
</div>
</div>
</div>
<div class="paragraph">
<p>You can add your custom profiles depending on the settings of your system and the workload manager you use (e.g. SGE, PBS, etc).</p>
</div>
<div class="paragraph">
<p>The line <code>clusterOptions</code> can be used to add any other option that you will usually use for your job submission.</p>
</div>
</div>
<div class="sect2">
<h3 id="_local_nextflow"><a class="link" href="#_local_nextflow">7.4. Local nextflow</a></h3>
<div class="paragraph">
<p>To avoid calling the pipeline using <code>./nextflow &#8230;&#8203;</code> you can modify the nextflow command like this <code>chmod 777 nextflow</code>. For running the pipeline you just need to use:</p>
</div>
<div class="sidebarblock">
<div class="content">
<div class="paragraph">
<p>nextflow run TransPi.nf &#8230;&#8203;</p>
</div>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_real_time_monitoring"><a class="link" href="#_real_time_monitoring">7.5. Real Time Monitoring</a></h3>
<div class="paragraph">
<p>To monitor your pipeline remotely without connecting to the server via ssh use <a href="https://tower.nf/login">Nextflow Tower</a>.
Make an account with your email and follow their instructions. After this, you can now run the pipeline adding the <code>-with-tower</code> option and follow live the execution
of the processes.</p>
</div>
<div class="sidebarblock">
<div class="content">
<div class="paragraph">
<p>nextflow run TransPi.nf --all -with-tower -profile conda</p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
Version 1.0.0<br>
Last updated 2021-02-18 11:11:15 -0800
</div>
</div>
</body>
</html>